dbms/src/DataStreams/SummingSortedBlockInputStream.cpp

```diff
@@ -349,13 +412,31 @@ bool SummingSortedBlockInputStream::mergeMap(const MapDescription & desc, Row &
 template <typename TSortCursor>
 bool SummingSortedBlockInputStream::addRow(Row & row, TSortCursor & cursor)
 {
-    bool res = mergeMaps(row, cursor);    /// Is there at least one non-zero number or non-empty array
-
-    for (size_t i = 0, size = column_numbers_to_sum.size(); i < size; ++i)
+    bool res = false;
+    for (auto & desc : columns_to_aggregate)
     {
-        size_t j = column_numbers_to_sum[i];
-        if (applyVisitor(FieldVisitorSum((*cursor->all_columns[j])[cursor->pos]), row[j]))
-            res = true;
+        if (desc.created)
+        {
+            // Specialized case for unary functions
+            if (desc.column_numbers.size() == 1)
+            {
+                auto & col = cursor->all_columns[desc.column_numbers[0]];
+                desc.function->add(desc.state.data(), &col, cursor->pos, nullptr);
+                // This stream discards rows that are zero across all summed columns
+                if (!res)
+                    res = col->get64(cursor->pos) != 0;
```

Comment:
It should discard rows, that was summed to zero (that becomes zero after summation).

Reply:
Right, added a better comment. For unsigned types, it's true that the row is summed to zero if all input numbers are also zero. The slight difference is with signed numbers sequence, like `-1 +1`, such row will result in zero, but it won't be deleted until the next merge. I don't know how to get state cheaply out of an aggregation function (there's only interface for inserting to column), which is why I used input.

Reply:
Ok.

(The only way to look at aggregation state is "insertResultInto" and it can be expensive.)